aiida/tools/importexport/migration/v03_to_v04.py

```diff
@@ -312,7 +312,8 @@ def _migration_calc_job_option_attribute_keys(attr_id, content):
                 'jobresource_params': 'resources',
                 'parser': 'parser_name'
             }
-            for key in content:
+            # Need to loop over a clone because the `content` needs to be modified in place
+            for key in copy.deepcopy(content):
```

Comment:
As a general solution I usually prefer to create a copy for the new, modified data instead of backing up the current data to be able to provide better guarantees in case of failure: the dict will always be in either the original unmodified or the fully processed state.
With a dict this could be done by using `content.clear() ; content.update(newdata)`.

Reply:
> As a general solution I usually prefer to create a copy for the new, modified data instead of backing up the current data to be able to provide better guarantees in case of failure

I agree, except here that was not really possible because of the code outside of this function. This code is in a function defined in a local scope and the caller reuses the `content` variable after the function is invoked. The changes therefore have to be made in place to the reference and I cannot return a copy. I didn't feel like rewriting the rest of the code, so I did this for now. Your preferred approach is exactly what I used in the other case though, because there it was possible.